### ISSUE TITLE
filterProtocolClaims deletes properties required by the IdTokenClaims type

### ISSUE THREAD

**Issue by niklasholm** [authts/oidc-client-ts#780]
If filterProtocolClaims is set to true (the default), the properties `iss`, `aud`, `exp` and `iat` that are **_required_** by the IdTokenClaims type are deleted and undefined, thus breaking the defined types.

```typescript
const userManager = new UserManager({
  // ... other options omitted
  filterProtocolClaims: true, // defaults to true
  // ...
});
const user = await userManager.getUser();
const exp = user.profile.exp; // TypeScript says this is a number
console.log(typeof exp) // Should be "number" but is "undefined"
```

**Relevant parts of the code**

These properties are deleted from the response:
https://github.com/authts/oidc-client-ts/blob/bcfe363c685e8f243b43bc588ecba7495f88ffa9/src/ResponseValidator.ts#L26-L41
https://github.com/authts/oidc-client-ts/blob/bcfe363c685e8f243b43bc588ecba7495f88ffa9/src/ResponseValidator.ts#L225-L235
Some of which are mandatory here:
https://github.com/authts/oidc-client-ts/blob/bcfe363c685e8f243b43bc588ecba7495f88ffa9/src/Claims.ts#L109

**Comment by pamapa** [comment]
Do you have a use case where you would need any of these properties?

The idea of the `ProtocolClaims` is to filter protocol details.
That said `iss` (issuer) and `aud` (audience) should probably **not** be filtered.

@brockallen Maybe we can drop the `ProtocolClaims` filter all together. Do you remember why you originally added it? I mean keeping some more properties in the profile should not hurt...

**Comment by brockallen** [comment]
Are those just claims added to the profile after the protocol handshake is complete? IIRC, it was just to redice the bloat of the storage "user" object in storage.

**Comment by brockallen** [comment]
The profile would be dynamic, by definition. This library doesn't know the claims expected based on the scopes, and there's no guarantee that all users have all claims.

As for the type definition defined, I don't know -- that's past my my era working on this library.

**Comment by pamapa** [comment]
@Badisi  I agree with you.

As such we have three valid options:
- A: drop the filtering completely
- B: do **not** filter "iss, sub, sub, aud, exp and iat" but do filter the remaining claims as is.
- C: make it configurable via settings, default would be B

We can process here step by step. We first implement option B, this still allows option A or C at a later stage.

@Badisi  Can you provide a merge request with option B?

**Comment by Badisi** [comment]
@pamapa, I went for C directly 😉

I didn't make any breaking change as I simply extended `filterProtocolClaims` to now also support a string array:
```ts
// before
filterProtocolClaims?: boolean;
// after
filterProtocolClaims?: boolean | string[];
```
---
Original issue reported by @niklasholm should be fixed as all required claims are now never filtered.
But it should be kept in mind that any other fields that are filtered will still appear as `undefined`.
I've tried updated the doc a little to reflect that.
Tests were also added to secure that feature.
